src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -249,11 +249,14 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
+            var pair = new KeyValuePair<Type, MethodInfo>();
+
             foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
                 if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    return xEquatableArgument.Value;
+                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
```

Comment:
There are two inheritance issues here:

1. The inheritance hierarchy of the class on which the interface is implemented

```C#
interface IFoo{}
class Foo : IFoo{}
class Foo2 : Foo, IFoo{}
```

We want to use the IFoo implementation in Foo2, when dealing with a Foo2 object. That's what @rprouse and I were refering to regarding the lack of guarantees in the return order from GetInterfaces. Since Microsoft doesn't guarantee the order, you would need to do this yourself.

2. The inheritance hierarchy of the Type parameters for the  interface. I think that's a subtle, difficult and highly postponable issue. :-)

The first issue is purely mechanical, although complicated, and I think it should be dealt with in the PR.